File: 2022/02/14/Q1 Two Sum.py
from typing import List

'''
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.'''


#
# def twoSum(nums: List[int], target: int) -> List[int]:
#     for i in range(0, len(nums)):
#         for j in range(i, len(nums)):
#             if nums[i] + nums[j] == target and i != j:
#                 return [i, j]
#     return []

# Faster solution:
def twoSum(nums: List[int], target: int) -> List[int]:
    check = {}

    for index, num in enumerate(nums):
        if target - num in check:
            return [index, check[target-num]]
        else:
            check[num] = index
    return []

print(twoSum([3, 2, 4], 6))

File: 2022/06/30/Stella/Q1_Two_Sum.cpp
// Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

// You may assume that each input would have exactly one solution, and you may not use the same element twice.

// You can return the answer in any order.
#include <iostream>
#include <string>
#include <map>
#include <unordered_map>
#include <algorithm>
#include <set>

using namespace std;


class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> numIndex;
        vector<int> result;
        for(int i=0; i<nums.size(); i++){
            if(numIndex.count(target-nums[i])){
                vector<int> sol = {i, numIndex[target-nums[i]]};
                return sol;
            }else{
                numIndex[nums[i]] = i;
            }
        }
        return result;
    }
};